File: LeetCode/198 - House Robber.cpp
// https://leetcode.com/problems/house-robber/submissions/

// Keep track of two numbers, including last element, and excluding last element, and build your current best sum based on that, and update for the next iteration.

class Solution {
public:
    int rob(vector<int>& nums) {
        if(nums.size() == 0)
            return 0;
        int inc, exc, cur_inc, cur_exc;
        inc = nums[0]; exc = 0;
        for(int i = 1; i < nums.size(); ++i){
            cur_exc = max(inc, exc);
            cur_inc = exc + nums[i];
            exc = cur_exc;
            inc = cur_inc;
        }
        return max(exc, inc);
    }
};

File: LeetCode/198 - House Robber.py
# https://leetcode.com/problems/house-robber/description/

class Solution:
    def rob(self, nums: List[int]) -> int:
        # dp[i] = Most money considering first i houses.
        # dp[i] = max(nums[i] + dp[i - 2], dp[i - 1])
        n = len(nums)
        cur = nums[0]
        excl_prev = 0
        for i in range(1, n):
            tmp = cur
            cur = max(nums[i] + excl_prev, cur)
            excl_prev = tmp
        return cur